### PR TITLE
Ruby: JWT Security Queries (CWE-347)

### PR DESCRIPTION
**Pull request by maikypedia**
This pull request adds `MissingJWTVerification` and `EmptyJWTSecret` queries covering `jwt` library.

Looking forward to your suggestions.

### CODE REVIEW

File: ruby/ql/src/experimental/cwe-347/EmptyJWTSecret.ql [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+ * @precision high
+ * @id rb/jwt-empty-secret-or-algorithm
+ * @tags security
+ */
+
+private import codeql.ruby.Concepts
+
+from JwtEncoding jwtEncoding
+where not jwtEncoding.signs()
+select jwtEncoding.getPayload(), "This JWT encoding uses empty key or none algorithm."
```
Comment (alexrford) [inline ruby/ql/src/experimental/cwe-347/EmptyJWTSecret.ql:15]:
```suggestion
select jwtEncoding.getPayload(), "This JWT encoding uses an empty key or none algorithm."
```

File: ruby/ql/src/experimental/cwe-347/EmptyJWTSecret.qhelp [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+<overview>
+<p>
+Applications encoding a JSON Web Token (JWT) may be vulnerable when it's not verified or algorithm is <code>none</code>.
+</p>
+</overview>
+
+<recommendation>
+<p>
+Use non-empty nor <code>None</code> values while encoding JWT payloads.
```
Comment (alexrford) [inline ruby/ql/src/experimental/cwe-347/EmptyJWTSecret.qhelp:14]: Nitpick, I found this line a little bit unclear. Maybe something like:
```suggestion
JSON Web Tokens should be signed using a strong cryptographic algorithm and non-empty secret.
```
I'm open to other suggestions!

**Review by alexrford — CHANGES_REQUESTED** [review]
Thanks for your contribution, this PR looks really good. Just a few nitpicky comments.

There's also a merge conflict in `Concepts.qll` and we should probably have a changenote to mention the two new experimental queries.

File: ruby/ql/lib/codeql/ruby/Concepts.qll [hunk 3]
```diff
@@ -1250,3 +1250,92 @@ module LdapExecution {
...
+  DataFlow::Node getPayload() { result = super.getPayload() }
+
+  /** Gets the argument containing the encoding algorithm. */
+  DataFlow::Node getAlgorithm() { result = super.getAlgorithm() }
+
+  /** Gets the argument containing the encoding key. */
+  DataFlow::Node getOptions() { result = super.getOptions() }
+
+  /** Checks if the signature gets verified while decoding. */
+  predicate verifies() { super.verifies() }
```
[the remark this answers is not in the document]
Comment (maikypedia) [reply ruby/ql/lib/codeql/ruby/Concepts.qll:1314]: Yes, `verifiesSignature()` would be clearer 😄

File: ruby/ql/src/experimental/cwe-347/EmptyJWTSecret.qhelp (hunk 2 above)
Comment (maikypedia) [reply ruby/ql/src/experimental/cwe-347/EmptyJWTSecret.qhelp:14]: 👍👍👍

**Review by alexrford — CHANGES_REQUESTED** [review]
@maikypedia sorry for the review delay, I was away for a while. I can merge this with this change-note modification and a resolution to the `Concepts.qll` merge conflict.

**Comment by maikypedia** [comment]
> @maikypedia sorry for the review delay, I was away for a while. I can merge this with this change-note modification and a resolution to the `Concepts.qll` merge conflict.

Np 🫡🫡🫡

**Review by alexrford — APPROVED** [review]
Thanks 👍

**Pull request merged**